### PR TITLE
pallet-xcm: ensure xcm outcome is always complete, revert effects otherwise

### LINKED ISSUE #2237: xcmPallet transfers should revert for incomplete XCM execution
Because it is silly to not do so and now we need to figure out how to rescue people's stuck funds https://kusama.subscan.io/extrinsic/20471678-2?event=20471678-36

The whole problem can be simplified if xtokens is used, which does the right thing https://github.com/open-web3-stack/open-runtime-module-library/blob/0abce5c717df8b47466ced37205fd961774a1eb2/xtokens/src/lib.rs#L669

### PR DESCRIPTION
**Pull request by xlc**
On extrinsics/call, ensure local XCM execution is complete/successful. Otherwise, fail the extrinsic so that state changes don't get committed to the db.

Added regression tests that fail without the fix.

fixes #2237

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: build-malus
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4398274

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-runtime-migration-rococo
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4398232

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-runtime-migration-collectives-westend
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4398238

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: build-test-parachain
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4398279

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4398225

File: polkadot/xcm/pallet-xcm/src/lib.rs [hunk 1]
```diff
@@ -1495,13 +1500,19 @@ impl<T: Config> Pallet<T> {
...
+		outcome.ensure_complete().map_err(|error| {
+			log::error!(target: "runtime::xcm", "XCM executio failed with error {:?}", error);
+			Error::<T>::LocalExecutionIncomplete
+		})?;
 
+		if let Some(remote_xcm) = remote_xcm {
 			let (ticket, price) = validate_send::<T::XcmRouter>(dest, remote_xcm.clone())
 				.map_err(Error::<T>::from)?;
 			if origin != Here.into_location() {
-				Self::charge_fees(origin, price).map_err(|_| Error::<T>::FeesNotMet)?;
```
Comment (xlc) [inline polkadot/xcm/pallet-xcm/src/lib.rs:1504]: someone should make a rule to disallow ignoring errors

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4398671

**Comment by xlc** [comment]
looks like I don't need to fix any existing tests? that means very bad test coverage.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4398817

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4399249

File: polkadot/xcm/pallet-xcm/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply polkadot/xcm/pallet-xcm/src/lib.rs:1504]: The error is clearly handled... What you want is "better observability".

Comment (xlc) [reply polkadot/xcm/pallet-xcm/src/lib.rs:1504]: yeah maybe I wasn’t using the right words but the original error details are ignored and that makes debugging hard. I have been bite by this kind of code too many times and I just don’t want to see it anymore

**Comment by xlc** [comment]
https://github.com/paritytech/polkadot-sdk/blob/1644e9285e017506e68cd7e9c30e466f66550493/polkadot/xcm/xcm-executor/integration-tests/src/lib.rs#L75 this test is got broken but it is writing in such way that I don't really know how to improve it other than rewrite it in a better way

I am busy and I don't want to spend any more time on this. so do whatever you want with this PR.

**Comment by acatangiu** [comment]
> I am busy and I don't want to spend any more time on this. so do whatever you want with this PR.

I can take it and drive to completion

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
This should be good to merge now, @KiChjang @franciscoaguirre ptal

**Review by franciscoaguirre — APPROVED** [review]

**Review by KiChjang — APPROVED** [review]

**Comment by xlc** [comment]
LGTM. Thanks for finishing it.

**Comment by bkchr** [comment]
@acatangiu this broke the short running benchmarks check in CI.

**Comment by acatangiu** [comment]
> @acatangiu this broke the short running benchmarks check in CI.

Looking into it

**Comment by bkchr** [comment]
Already have a fix.

**Comment by acatangiu** [comment]
Let's see if this fixes it https://github.com/paritytech/polkadot-sdk/pull/2489

**Comment by xlc** [comment]
This is not just about observability of errors. The important thing is no partial local XCM execution @KiChjang

**Comment by acatangiu** [comment]
@KiChjang afaik extrinsics are by default using [transactional storage](https://github.com/paritytech/polkadot-sdk/blob/bc53b9a03a742f8b658806a01a7bf853cb9a86cd/substrate/frame/support/src/storage/transactional.rs) - so that nothing the extrinsic does (in regards to storage/state at least) gets committed to lower layers (DB) unless the extrinsic returns `Ok`, aka _full_ extrinsic executes to completion.

So the main problem this PR fixes is to protect against partial XCM execution side-effects, not just better observability of errors.
By returning error on `outcome != complete`, all storage/state changes done by XCM instructions executed so far, are discarded; extrinsic call has noop effects on state.

Extra care needs to be taken for side-effects not regarding state (like sending outbound XCMs) - which is also covered correctly here afaict.

**Pull request merged**